Add more bindings to scroll-transient-state

Somelauw opened the pull request:
Add more bindings to scroll-transient-state

As described in #7995. Once in scroll-transient-state more bindings are available. I haven't added new SPC n mappings for scrolling yet. Let me know if I should add these too or if perhaps there should be a generic binding that switches to scroll-transient-state/body.

Edit: Originally, I had 0 and $ bound to evil-scroll-left and evil-scroll-right (just like H and L). I have removed them, because these keys are probably expected to scroll to the very first and last column of the buffer.

deb0ch commented:
I like this PR, because the bindings are consistent with what you would expect of Evil motions :smile_cat: 

It would be very instinctive I think to activate the transient state on any of these bindings, like `SPC n j`, `SPC n $`, etc... When you do one scroll command, isn't it likely that you will want to do many to move around your buffer ?

It would conflict with the narrowing stuff though, but the `SPC N` prefix is free so we could move narrowing to that :smile_cat: 

But now that I think of it, why are the **s**crolling actions under `SPC n` in the first place ?

@syl20bnr It seems that you are the master of key bindings, any insights ? :smiley_cat:

bmag commented:
This looks nice. Can we put narrowing on `SPC n` and scrolling on `SPC N`? Reasons:
- narrowing is quite useful at times
- there are already easy bindings for most scrolling commands
- the TS lowers the inconvenience of having scrolling on `SPC N`

deb0ch commented:
I don't think it's a bad thing or that it should be changed, but the question remains: why are the scrolling actions under the letter `n` or `N` ?

bmag commented:
@deb0ch I don't know, maybe because these are "navigation" commands (and "s" is taken by "search")

deb0ch commented:
Oooooooh THAT'S the mnemonic :open_mouth: 

Thanks @bmag ! :smiley_cat: :+1:

Somelauw commented:
I have put "narrow" back to n and moved "navigation" to N now. I have also squashed the commits.
Should I keep the original "N,", "N.", "N<" and "N>" shortcuts?

bmag commented:
@Somelauw just tried it, and it's very convenient. I think we can improve it with:
- `<`, `>`: go to start/end of buffer (evil-goto-first-line`, `evil-goto-line`) (that will also resemble emacs style `M-<` and `M->`)
- `,`, `.`: they don't do any harm as they are, but maybe we can bind `,` to `evil-jump-backward` and `.` to `evil-jump-forward` (that will also resemble emacs style `M-.` and `M-,`)

If no one has any objections, I'll merge it once @Somelauw applies these changes.

Somelauw commented:
I have added the requested shortcuts. In addition, I have added a hydra-docstring for readability.

bmag commented:
Hmm.. for some reason `SPC N .` (spacemacs/scroll-transient-state/evil-jump-forward) doesn't move the cursor (`SPC N ,` works fine). Do you experience the same problem?

Somelauw commented:
> Hmm.. for some reason SPC N . (spacemacs/scroll-transient-state/evil-jump-forward) doesn't move the cursor (SPC N , works fine). Do you experience the same problem?

@bmag Thanks. I have noticed that both evil-jump-forward and evil-jump-backward behave differently in scroll-transient-state:

- evil-jump-backward (,)
Behaviour: goes back one position on its first invocation but subsequent key presses jump back and forth.
Expected: Pressing it multiple times goes further back in history

- evil-jump-forward (.)
Behaviour: Does nothing
Expected: Undoes evil-jump-backward

When using evil's built-in C-o and C-i key bindings, these commands behave as expected. I'm not sure what causes them to behave differently in scroll-transient-state.

bmag commented:
@Somelauw so lets drop the `,` and `.` bindings and get this merged?

Somelauw commented:
I have dropped them for now.

bmag commented:
Thank you :+1: Cherry-picked into develop, you can safely delete your branch.
